### PR TITLE
Planning: add soft s bound in nonlinear speed opt

### PR DESCRIPTION
**Pull request by JasonZhou404**

### CODE REVIEW

File: modules/planning/tasks/optimizers/piecewise_jerk_speed/piecewise_jerk_speed_nonlinear_ipopt_interface.cc [hunk 1]
```diff
@@ -21,32 +21,48 @@
...
     int &n, int &m, int &nnz_jac_g, int &nnz_h_lag,
     IndexStyleEnum &index_style) {
   num_of_variables_ = num_of_points_ * 3;
+
+  if (use_soft_safety_bound_) {
+    // complementary slack variable for soft upper and lower s bound
+    num_of_variables_ += num_of_points_ * 2;
+
+    lower_s_slack_offset_ = num_of_points_ * 3;
+
```
Comment (xiaoxq) [inline modules/planning/tasks/optimizers/piecewise_jerk_speed/piecewise_jerk_speed_nonlinear_ipopt_interface.cc]: Could you make lines more compact?

File: modules/planning/tasks/optimizers/piecewise_jerk_speed/piecewise_jerk_speed_nonlinear_ipopt_interface.cc [hunk 2]
```diff
@@ -570,6 +726,31 @@ bool PiecewiseJerkSpeedNonlinearIpoptInterface::eval_jac_g(
...
+      // s_i - soft_upper_s_i - upper_slack_i <= 0.0
+      for (int i = 0; i < num_of_points_; ++i) {
+        // s_i
+        values[non_zero_index] = 1.0;
+        non_zero_index++;
+
+        // upper_slack_i
+        values[non_zero_index] = -1.0;
+        non_zero_index++;
```
Comment (xiaoxq) [inline modules/planning/tasks/optimizers/piecewise_jerk_speed/piecewise_jerk_speed_nonlinear_ipopt_interface.cc]: `++a` is more efficient.

File: modules/planning/tasks/optimizers/piecewise_jerk_speed/piecewise_jerk_speed_nonlinear_ipopt_interface.h [hunk 3]
```diff
@@ -180,6 +192,8 @@ class PiecewiseJerkSpeedNonlinearIpoptInterface : public Ipopt::TNLP {
 
   double w_overall_centripetal_acc_ = 500.0;
 
+  double w_soft_s_bound_ = 0.0;
+
```
Comment (xiaoxq) [inline modules/planning/tasks/optimizers/piecewise_jerk_speed/piecewise_jerk_speed_nonlinear_ipopt_interface.h]: Ditto. More compact?

**Comment by qwetqwe** [comment]
Hi, @JasonZhou404 , I want to know more about slack variable of soft bounds. Why do you add slack variables to state variable x[i]? 
s_i - soft_upper_s_i - upper_slack_i <= 0.0
s_i - soft_lower_s_i + lower_slack_i >= 0.0
upper_slack_i>=0
lower_slack_i>=0
I think it is equal to:
s_i - soft_upper_s_i <=0
s_i - soft_lower_s_i >=0
Is it right?

**Pull request merged**